randgen: disable generation of REGNAMESPACE type expressions

msirek opened the pull request:
Casting an OID to REGNAMESPACE runs this SQL using the internal executor:
```
SELECT pg_namespace.oid, nspname FROM pg_catalog.pg_namespace WHERE oid = $1
```
When the `GenerateConstrainedScans` rule is disabled, this returns no
rows and so doesn't apply the cast. The same SQL, not run via an
internal executor, but also with `GenerateConstrainedScans` disabled,
behaves correctly. Disabling `GenerateConstrainedScans` all the time
prevents the cluster from starting.

The reason a full scan of pg_namespace doesn't find the OID is because it
only checks for schemas in the current database:
https://github.com/cockroachdb/cockroach/blob/7b341ffa678e4a22416e2274351fd56f415f5421/pkg/sql/virtual_schema.go#L602
https://github.com/cockroachdb/cockroach/blob/d10c3dd42c3dc40cad82792a30ae47fd2a663f43/pkg/sql/pg_catalog.go#L2099-L2106
https://github.com/cockroachdb/cockroach/blob/a7e9c4a68b81436d1f9382518d4267f74cbdac94/pkg/sql/information_schema.go#L2295-L2296

Whereas with a constrained scan, all descriptors are searched:
https://github.com/cockroachdb/cockroach/blob/af6a72a622ae05f3733b5db637403b3eaa9455f1/pkg/sql/catalog/descs/descriptor.go#L168-L175

The correct result is from the constrained scan, because we currently
allow cross-database references. Once #55791 is fixed, both results
should match.

The fix alternatives are:
1. disallow disabling of the `GenerateConstrainedScans` rule for internal SQL
2. turn off generation of REGNAMESPACE expressions in randgen to avoid hitting this problem in tests.

The 2nd fix alternative is implemented to avoid losing any of the
rule-disabling test coverage provided by the
`testing_optimizer_disable_rule_probability` setting.

Fixes #98322

stevendanna commented on pkg/sql/internal.go:
```diff
@@ -709,13 +709,21 @@ func (ie *InternalExecutor) QueryIteratorEx(
+	// which may rely on rewrite rules for proper behavior, could potentially make
+	// the cluster unstable and may make randomized tests more noisy with
+	// difficult to debug behavior.
+	if !forStmtBundle {
```
Just to make sure I'm following: We exclude the statement bundle here because generating the statement bundle runs the user-query through the internal executor and we want it to behave the same?

stevendanna reviewed (commented):
I don't have much expertise on the underlying casting issue, but the explanation on why we would want to exclude this given here makes sense to me.

stevendanna commented:
Does this need `backport-23.1` as well?

msirek reviewed (commented):
Thanks, added the backport label.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/99168#-:-NR7yPeg3Akmz7A9QK4S:b-q74b0f)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach, @mgartner, and @stevendanna)

___
*[`pkg/sql/internal.go` line 724 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/99168#-NR5BF9W8Nx0ATbLqt32-r2-724:-NR7yUJc6WgAp4uIq_bG:b-gh7115) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5833cc9da5e6146381a15f917179b6f4e17c112/pkg/sql/internal.go#L724)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Just to make sure I'm following: We exclude the statement bundle here because generating the statement bundle runs the user-query through the internal executor and we want it to behave the same?
</blockquote></details>

That's correct, specifically statement bundle generation issues SHOW statements on each of the session flags via an internal executor, so it relies on those flags matching the original values.  The `TestExplainAnalyzeDebug` test fails otherwise.
___
*[`pkg/sql/internal.go` line 805 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/99168#-NR5BF9W8Nx0ATbLqt32-r2-805:-NR7z2fz449G2PCcSjKS:b-wu4tz3) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5833cc9da5e6146381a15f917179b6f4e17c112/pkg/sql/internal.go#L805)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Looks like this was maybe left behind from some debugging, or is this doing something I don't understand?
</blockquote></details>

Thanks, removed it.
___
*[`pkg/sql/internal.go` line 810 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/99168#-NR5BF9W8Nx0ATbLqt32-r2-810:-NR7zgpX7fFmxoiDzQ5G:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5833cc9da5e6146381a15f917179b6f4e17c112/pkg/sql/internal.go#L810)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

[nit] We could make"stmtEnvCollector" to a constant in explain_bundle.go
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->

cucaroach commented:
Why does it not run correctly if `GenerateConstrainedScans` is disabled?  Should we worry about disabling this rule causing other correctness noise?  Did we think about making this rule a special rule that can't be disabled?  I'm having trouble swallowing the "we can never do optimizer perturbation testing on internal queries" pill although as a temporary measure I'm fine with it.

msirek reviewed (commented):
The rationale for this approach is server stability over test coverage. I suppose if we never explicitly set this flag for internal SQL, it's still relatively safe. OK, I'll do a bit more digging to see if we don't have to give up any test coverage.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/99168#-:-NR897QA3md1NbP_65XR:bygb8s2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach, @mgartner, and @stevendanna)



<!-- Sent from Reviewable.io -->

msirek reviewed (commented):
> Maybe the safe expedient thing to do is just make this rule non-optional until we understand the pathology?

Non-optional for all SQLs, or just internal SQLs? If for all SQLs, then we'd be losing test coverage for regular, non-internal statements.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/99168#-:-NR8dkcA-Aq0djKLdw5x:b-tkho4o)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach, @mgartner, and @stevendanna)



<!-- Sent from Reviewable.io -->

msirek reviewed (commented):
The reason a full scan of pg_namespace doesn't find the OID is because it only checks for schemas in the current database:
https://github.com/cockroachdb/cockroach/blob/7b341ffa678e4a22416e2274351fd56f415f5421/pkg/sql/virtual_schema.go#L602
https://github.com/cockroachdb/cockroach/blob/d10c3dd42c3dc40cad82792a30ae47fd2a663f43/pkg/sql/pg_catalog.go#L2099-L2106
https://github.com/cockroachdb/cockroach/blob/a7e9c4a68b81436d1f9382518d4267f74cbdac94/pkg/sql/information_schema.go#L2295-L2296

Whereas with a constrained scan, all descriptors are searched:
https://github.com/cockroachdb/cockroach/blob/af6a72a622ae05f3733b5db637403b3eaa9455f1/pkg/sql/catalog/descs/descriptor.go#L168-L175
```
descs.(*byIDLookupContext).lookupLeased (descriptor.go:377) github.com/cockroachdb/cockroach/pkg/sql/catalog/descs
<autogenerated>:2
descs.getDescriptorsByID (descriptor.go:180) github.com/cockroachdb/cockroach/pkg/sql/catalog/descs
descs.ByIDGetter.Desc (getters.go:48) github.com/cockroachdb/cockroach/pkg/sql/catalog/descs
descs.ByIDGetter.Schema (getters.go:74) github.com/cockroachdb/cockroach/pkg/sql/catalog/descs
sql.glob..func241.1 (pg_catalog.go:2142) github.com/cockroachdb/cockroach/pkg/sql
sql.glob..func241 (pg_catalog.go:2162) github.com/cockroachdb/cockroach/pkg/sql
sql.(*virtualDefEntry).makeConstrainedRowsGenerator.func1 (virtual_schema.go:787) github.com/cockroachdb/cockroach/pkg/sql
sql.setupGenerator.func3 (virtual_table.go:127) github.com/cockroachdb/cockroach/pkg/sql
stop.(*Stopper).RunAsyncTaskEx.func2 (stopper.go:470) github.com/cockroachdb/cockroach/pkg/util/stop
runtime.goexit (asm_amd64.s:1594) runtime
 - Async Stack Trace
stop.(*Stopper).RunAsyncTaskEx (stopper.go:461) github.com/cockroachdb/cockroach/pkg/util/stop
```

I suppose this could be fixed by using a nil database descriptor in the full scan case, but maybe this could break other assumptions in the code.  At the very least it increases the processing path.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/99168#-:-NR909x2DnMGePOsH0PF:b-87o9xc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach, @mgartner, and @stevendanna)



<!-- Sent from Reviewable.io -->

msirek reviewed (commented):
Question for schema team: is it expected that an unconstrained scan from a virtual table should read all rows (descriptors) from all of the same sources as a constrained scan?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/99168#-:-NR94RsyA4xSpJ9bBFm1:bghrr11)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach, @mgartner, and @stevendanna)



<!-- Sent from Reviewable.io -->

msirek reviewed (commented):
Taking a look at postgres behavior, it looks like they do not allow a cast to REGNAMESPACE for objects not belonging to the current database:
```
postgres=# SELECT 49433::REGNAMESPACE;
 regnamespace 
--------------
 mws
(1 row)

postgres=# SELECT pg_namespace.oid, nspname FROM pg_catalog.pg_namespace;
  oid  |      nspname       
-------+--------------------
    99 | pg_toast
    11 | pg_catalog
  2200 | public
 12432 | information_schema
 49433 | mws
(5 rows)

postgres=# \c mws2;
psql (12.11 (Ubuntu 12.11-0ubuntu0.20.04.1), server 15devel)
WARNING: psql major version 12, server major version 15.
         Some psql features might not work.
You are now connected to database "mws2" as user "postgres".
mws2=# SELECT 49433::REGNAMESPACE;
 regnamespace 
--------------
 49433
(1 row)

mws2=# SELECT pg_namespace.oid, nspname FROM pg_catalog.pg_namespace;
  oid  |      nspname       
-------+--------------------
    99 | pg_toast
    11 | pg_catalog
  2200 | public
 12432 | information_schema
(4 rows)

```
So, the results of the constrained scan may actually be incorrect, and the unconstrained scan is correct.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/99168#-:-NR97orxBxrreSOE8PL9:bngssu0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach, @mgartner, and @stevendanna)



<!-- Sent from Reviewable.io -->

msirek reviewed (commented):
Some more discussion is here: https://cockroachlabs.slack.com/archives/CARM0PTK4/p1679504546690839

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/99168#-:-NR9NBmp5HlM3fAElHGH:b4i7jb7)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @cucaroach, @mgartner, and @stevendanna)



<!-- Sent from Reviewable.io -->

chengxiong-ruan commented:
As I chatted with @msirek that the main issue here is that corss db reference is not disallowed when looking up objects with OIDs. I think disallowing it now is a bit risky since we've been allowing it all the time. I agree that we should just disable the constraint scan rule for now to get this test coverage green. Allowing cross-db reference seems a bigger issue worth more discussions unfortunately.

rafiss commented:
I agree. For now, let's still allow the cross-DB reference. https://github.com/cockroachdb/cockroach/issues/55791 is the issue that tracks disallowing cross-DB references.

mgartner commented on pkg/sql/internal.go:
```diff
@@ -709,13 +709,21 @@ func (ie *InternalExecutor) QueryIteratorEx(
+	// which may rely on rewrite rules for proper behavior, could potentially make
+	// the cluster unstable and may make randomized tests more noisy with
+	// difficult to debug behavior.
+	if !forStmtBundle {
```
I don't understand the name `forStmtBundle` here. What do statement bundles have to do with disabling rules in the internal executor?

msirek replied:
> I don't understand the name forStmtBundle here. What do statement bundles have to do with disabling rules in the internal executor?

Statement bundle collection issues a number of internal `SET` statements to gather session setting values.
If this case weren't excluded, it would return the wrong value for `testing_optimizer_disable_rule_probability`.

mgartner reviewed (commented):
We can leave the cross-DB references as-is for now, but I'd like to find a way to silence this failure. Is this the best way? Should we just disable REGNAMESPACE-related expressions in sqlsmith/randgen (not sure that would solve this entirely)?

msirek reviewed (commented):
The current fix disallows disabling all rewrites for internal SQL, but I'm thinking of changing it to just disallow disabling of `GenerateConstrainedScans` for internal SQL.  Disabling REGNAMESPACE and maybe other related types in sqlsmith/randgen is definitely another option. Let me see what other types might need to be handled.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/99168#-:-NRAMB2-4mwNIjJWlBm7:b-4s04tj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @cucaroach, and @stevendanna)



<!-- Sent from Reviewable.io -->

Commit:
randgen: disable generation of REGNAMESPACE type expressions

Casting an OID to REGNAMESPACE runs this SQL using the internal executor:
```
SELECT pg_namespace.oid, nspname FROM pg_catalog.pg_namespace WHERE oid = $1
```
When the `GenerateConstrainedScans` rule is disabled, this returns no
rows and so doesn't apply the cast. The same SQL, not run via an
internal executor, but also with `GenerateConstrainedScans` disabled,
behaves correctly. Disabling `GenerateConstrainedScans` all the time
prevents the cluster from starting.

The reason a full scan of pg_namespace doesn't find the OID is because it
only checks for schemas in the current database:
https://github.com/cockroachdb/cockroach/blob/7b341ffa678e4a22416e2274351fd56f415f5421/pkg/sql/virtual_schema.go#L602
https://github.com/cockroachdb/cockroach/blob/d10c3dd42c3dc40cad82792a30ae47fd2a663f43/pkg/sql/pg_catalog.go#L2099-L2106
https://github.com/cockroachdb/cockroach/blob/a7e9c4a68b81436d1f9382518d4267f74cbdac94/pkg/sql/information_schema.go#L2295-L2296

Whereas with a constrained scan, all descriptors are searched:
https://github.com/cockroachdb/cockroach/blob/af6a72a622ae05f3733b5db637403b3eaa9455f1/pkg/sql/catalog/descs/descriptor.go#L168-L175

The correct result is from the constrained scan, because we currently
allow cross-database references. Once #55791 is fixed, both results
should match.

The fix alternatives are:
1. disallow disabling of the `GenerateConstrainedScans` rule for internal SQL
2. turn off generation of REGNAMESPACE expressions in randgen to avoid hitting this problem in tests.

The 2nd fix alternative is implemented to avoid losing any of the
rule-disabling test coverage provided by the
`testing_optimizer_disable_rule_probability` setting.

Fixes #98322

Release note: None

msirek reviewed (commented):
OK, I went with the approach of disabling generation of REGNAMESPACE type expressions in randgen.
RFAL

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/99168#-:-NRE8fqk07MRBHBT6Lhh:b-xcvi5p)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @cucaroach, @mgartner, and @stevendanna)



<!-- Sent from Reviewable.io -->

cucaroach reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Thanks for seeing this one through!

msirek commented:
TFTRs!
bors r+
